File: src/bin/enum/enum_.py
import dataclasses
import typing

PlaylistId = typing.NewType("PlaylistId", int)


@dataclasses.dataclass
class Start:
    playlist_id: PlaylistId


@dataclasses.dataclass
class Pause:
    pass


@dataclasses.dataclass
class Stop:
    pass


UserAction = Start | Pause | Stop


def main() -> None:
    # [...]
    action = Start(42)
    # [...]

    match action:
        case Start(playlist_id):
            print(f"Start playlist {playlist_id}")
        case Pause():
            print("Pause")
        case Stop():
            print("Stop")
        case _:
            typing.assert_never(action)


if __name__ == "__main__":
    main()

File: src/bin/enum/enum.rs
pub type PlaylistId = i32;

pub enum UserAction {
    Start(PlaylistId),
    Pause,
    Stop,
}

pub fn main() {
    // [...]
    let action = UserAction::Start(42);
    // [...]

    match action {
        UserAction::Start(playlist_id) => {
            println!("Start playlist {}", playlist_id)
        }
        UserAction::Pause => {
            println!("Pause")
        }
        UserAction::Stop => {
            println!("Stop")
        }
    }
}
